File: minidapp/hosting.js
/* Hosting (MiniDapp) — upload chosen photos to the user's own storage and
 * hand back the public URLs the mint seals. Browser side of the Android
 * Hosting.java: the pure helpers (slug / fillTemplate / publicUrl /
 * parseKuboAdd) are byte-parity with the Java, pinned by the SHARED fixtures
 * in test/fixtures/hosting/parity.json. Transport is browser fetch() —
 * MDS.net.POST is string-only, no good for binary — so the HTTP backends are
 * WebDAV, kubo, Pinata and GitHub (all CORS-reachable when the user enables
 * it). SFTP is Android-only; hidden here. Verification uses MDS.net.GET
 * (node-side, no CORS). Profiles live in MDS.keypair (on-node, per-dapp). */
"use strict";

var HOSTING = (function () {

  var TYPES = ["webdav", "kubo", "pinata", "github"];   // no sftp in the browser

  /* ---- storage (MDS keypair) ---- */

  function load(cb) {
    MDS.keypair.get("hosting_profiles", function (r) {
      var arr = [];
      try {
        if (r && r.status && r.response && r.response.value) {
          arr = JSON.parse(decodeURIComponent(r.response.value));
        }
      } catch (e) { arr = []; }
      cb(Array.isArray(arr) ? arr : []);
    });
  }

  function save(profiles, cb) {
    MDS.keypair.set("hosting_profiles", encodeURIComponent(JSON.stringify(profiles)),
      function (r) { if (cb) cb(r && r.status); });
  }

  function getDefault(profiles) {
    for (var i = 0; i < profiles.length; i++) if (profiles[i].def) return profiles[i];
    return profiles.length ? profiles[0] : null;
  }

  /* ---- pure helpers (parity with Hosting.java) ---- */

  function slug(s) {
    if (s == null) s = "";
    var out = ("" + s).normalize("NFD").replace(/[̀-ͯ]/g, "")
      .toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/-+/g, "-").replace(/^-|-$/g, "");
    if (out.length > 40) out = out.substring(0, 40).replace(/-$/, "");
    return out === "" ? "x" : out;
  }

  function fillTemplate(tpl, tokens) {
    var out = tpl == null ? "" : tpl;
    for (var k in tokens) {
      if (tokens.hasOwnProperty(k)) {
        out = out.split("{" + k + "}").join(tokens[k] == null ? "" : tokens[k]);
      }
    }
    return out;
  }

  function ts36() { return Math.floor(Date.now() / 1000).toString(36); }
  function rand4() { return (0x10000 + Math.floor(Math.random() * 0xFFFF)).toString(16).substring(1); }

  function trimSlash(s) {
    if (!s) return "";
    return s.charAt(s.length - 1) === "/" ? s.substring(0, s.length - 1) : s;
  }

  function cfgOf(p) { return p[p.type] || {}; }

  function publicUrl(p, relPathOrCid, isDir) {
    var c = cfgOf(p), type = p.type, tail = isDir ? "/" : "";
    if (type === "kubo" || type === "pinata") {
      var gw = trimSlash(c.gateway || "");
      if (!gw && type === "pinata") gw = "https://gateway.pinata.cloud";
      return gw + "/ipfs/" + relPathOrCid + tail;
    }
    if (type === "github") {
      if (c.serve === "pages") return trimSlash(c.pagesPrefix || "") + "/" + relPathOrCid + tail;
      return "https://raw.githubusercontent.com/" + c.owner + "/" + c.repo
        + "/" + (c.branch || "main") + "/" + relPathOrCid + tail;
    }
    return trimSlash(c.urlPrefix || "") + "/" + relPathOrCid + tail;
  }

  function parseKuboAdd(ndjson, dirName) {
    if (!ndjson) return "";
    var lines = ndjson.split("\n"), last = "";
    for (var i = 0; i < lines.length; i++) {
      var line = lines[i].trim();
      if (!line) continue;
      try {
        var o = JSON.parse(line);
        if (o.Name === dirName) return o.Hash || "";
        if (o.Hash) last = o.Hash;
      } catch (e) { }
    }
    return (dirName == null || dirName === "") ? last : "";
  }

  /* ---- transport (browser fetch) ---- */

  function b64(bytes, cb) {
    var blob = bytes instanceof Blob ? bytes : new Blob([bytes]);
    var r = new FileReader();
    r.onload = function () { cb(("" + r.result).split(",")[1]); };
    r.readAsDataURL(blob);
  }

  // WebDAV PUT (Basic auth). files: [{blob, rel, mime}]
  function putWebdav(p, files, done) {
    var c = cfgOf(p), endpoint = trimSlash(c.endpoint || "");
    var auth = c.user ? ("Basic " + btoa(c.user + ":" + (c.password || ""))) : null;
    var i = 0, base = "";
    (function next() {
      if (i >= files.length) { done(null, files.length === 1 ? publicUrl(p, files[0].rel, false) : base); return; }
      var f = files[i];
      var headers = { "Content-Type": f.mime || "application/octet-stream" };
      if (auth) headers.Authorization = auth;
      fetch(endpoint + "/" + f.rel, { method: "PUT", headers: headers, body: f.blob })
        .then(function (res) {
          if (!res.ok) { done("WebDAV PUT failed (HTTP " + res.status + ") at " + f.rel); return; }
          base = publicUrl(p, f.rel.replace(/[^/]*$/, ""), false);
          i++; next();
        })
        .catch(function (e) { done("WebDAV unreachable — " + e + " (check CORS on your server)"); });
    })();
  }

  // kubo /api/v0/add — one multipart with N parts under "dir/"
  function putKubo(p, files, isDir, done) {
    var c = cfgOf(p);
    var fd = new FormData();
    for (var i = 0; i < files.length; i++) {
      fd.append("file", files[i].blob, (isDir ? "dir/" : "") + files[i].name);
    }
    var pin = c.pin === false ? "false" : "true";
    fetch(trimSlash(c.apiUrl || "") + "/api/v0/add?pin=" + pin + "&cid-version=1&progress=false",
      { method: "POST", body: fd })
      .then(function (res) { return res.text(); })
      .then(function (text) {
        var cid = parseKuboAdd(text, isDir ? "dir" : "");
        if (!cid) { done("kubo add returned no CID"); return; }
        done(null, publicUrl(p, cid, isDir));
      })
      .catch(function (e) { done("kubo unreachable — " + e + " (enable CORS on the node: ipfs config API.HTTPHeaders.Access-Control-Allow-Origin)"); });
  }

  function putPinata(p, files, isDir, done) {
    var c = cfgOf(p);
    if (!c.jwt) { done("No Pinata JWT in the profile"); return; }
    var fd = new FormData();
    for (var i = 0; i < files.length; i++) {
      fd.append("file", files[i].blob, (isDir ? "dir/" : "") + files[i].name);
    }
    fetch("https://api.pinata.cloud/pinning/pinFileToIPFS",
      { method: "POST", headers: { Authorization: "Bearer " + c.jwt }, body: fd })
      .then(function (res) { return res.json().then(function (j) { return { ok: res.ok, j: j, status: res.status }; }); })
      .then(function (r) {
        if (!r.ok) { done("Pinata pin failed (HTTP " + r.status + ")" + (r.status === 401 ? " — check the JWT" : "")); return; }
        if (!r.j.IpfsHash) { done("Pinata returned no CID"); return; }
        done(null, publicUrl(p, r.j.IpfsHash, isDir));
      })
      .catch(function (e) { done("Pinata unreachable — " + e); });
  }

  // GitHub contents API PUT (JSON b64). No sha -> 422 on overwrite.
  function putGithub(p, files, done) {
    var c = cfgOf(p), i = 0, base = "";
    (function next() {
      if (i >= files.length) { done(null, files.length === 1 ? publicUrl(p, files[0].rel, false) : base); return; }
      var f = files[i];
      b64(f.blob, function (content) {
        fetch("https://api.github.com/repos/" + c.owner + "/" + c.repo + "/contents/" + f.rel,
          { method: "PUT",
            headers: { Authorization: "Bearer " + c.token, Accept: "application/vnd.github+json",
                       "X-GitHub-Api-Version": "2022-11-28", "Content-Type": "application/json" },
            body: JSON.stringify({ message: "atelier: " + f.rel, content: content, branch: c.branch || "main" }) })
          .then(function (res) {
            if (res.status === 422) { done("GitHub refused: " + f.rel + " already exists (never overwritten)"); return; }
            if (res.status === 401 || res.status === 403) { done("GitHub auth failed (HTTP " + res.status + ") — check the token"); return; }
            if (!res.ok) { done("GitHub upload failed (HTTP " + res.status + ") at " + f.rel); return; }
            base = publicUrl(p, f.rel.replace(/[^/]*$/, ""), false);
            i++; next();
          })
          .catch(function (e) { done("GitHub unreachable — " + e); });
      });
    })();
  }

  /* ---- orchestration ---- */

  // single file -> {blob, mime, ext}; lane names the field
  function uploadSingle(p, blob, mime, ext, collSlug, laneName, done) {
    var rel = fillTemplate(p.pathTemplate || "atelier/{collection}/{name}-{ts}{ext}", {
      collection: collSlug, name: laneName, ext: ext, ts: ts36(), rand: rand4(), idx: ""
    });
    var files = [{ blob: blob, rel: rel, name: rel.replace(/^.*\//, ""), mime: mime }];
    dispatch(p, files, false, function (err, url) { done(err, url); });
  }

  // N plates -> base + ext
  function uploadPlates(p, plates, collSlug, done) {
    // plates: [{blob, mime, ext}] in order; homogeneous ext enforced by caller
    var dirRel = fillTemplate(p.dirTemplate || "atelier/{collection}-{ts}", {
      collection: collSlug, ts: ts36(), rand: rand4()
    });
    var files = [];
    for (var i = 0; i < plates.length; i++) {
      files.push({ blob: plates[i].blob, rel: dirRel + "/" + (i + 1) + plates[i].ext,
                   name: (i + 1) + plates[i].ext, mime: plates[i].mime });
    }
    dispatch(p, files, true, function (err, base) { done(err, base); });
  }

  function dispatch(p, files, isDir, done) {
    switch (p.type) {
      case "webdav": putWebdav(p, files, done); break;
      case "kubo":   putKubo(p, files, isDir, done); break;
      case "pinata": putPinata(p, files, isDir, done); break;
      case "github": putGithub(p, files, done); break;
      default: done("This destination type isn't available in the browser (SFTP is Android-only)");
    }
  }

  /* ---- verification (node-side GET, no CORS) ---- */

  function verify(urls, done) {
    var i = 0;
    (function next() {
      if (i >= urls.length) { done(null); return; }
      MDS.net.GET(urls[i], function (res) {
        if (!res || !res.status || !res.response || !res.response.web) {
          done("Hosted URL does not serve: " + urls[i]); return;
        }
        i++; next();
      });
    })();
  }

  function testProfile(p, done) {
    var nonce = "atelier-" + Date.now();
    var blob = new Blob([nonce], { type: "text/plain" });
    var rel = fillTemplate("atelier/probe/{ts}-{rand}.txt", { ts: ts36(), rand: rand4() });
    var files = [{ blob: blob, rel: rel, name: rel.replace(/^.*\//, ""), mime: "text/plain" }];
    dispatch(p, files, false, function (err, url) {
      if (err) { done("failed: " + err); return; }
      verify([url], function (verr) {
        done(verr ? ("uploaded ✓ but fetch failed: " + verr) : ("uploaded ✓ · fetched ✓\n" + url));
      });
    });
  }

  return {
    TYPES: TYPES, load: load, save: save, getDefault: getDefault,
    slug: slug, fillTemplate: fillTemplate, publicUrl: publicUrl, parseKuboAdd: parseKuboAdd,
    ts36: ts36, rand4: rand4,
    uploadSingle: uploadSingle, uploadPlates: uploadPlates, verify: verify, testProfile: testProfile
  };
})();

File: android/app/src/main/java/com/eurobuddha/statenft/Hosting.java
package com.eurobuddha.statenft;

import org.json.JSONArray;
import org.json.JSONObject;

import java.net.HttpURLConnection;
import java.net.URL;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/** Hosting — upload chosen photos to the USER'S OWN storage and hand back the
 *  public URLs the mint seals on-chain. One profile schema, six backends
 *  (SFTP / WebDAV / kubo IPFS / Pinata / GitHub / Arweave), one law: what the
 *  mint writes must be a URL that provably serves, and no remote file is ever
 *  overwritten. Pure helpers are static and mirrored byte-for-byte in the
 *  MiniDapp's hosting.js — parity pinned by shared fixtures under
 *  test/fixtures/hosting/. Arweave is Android-only (no hosting.js counterpart). */
final class Hosting {

    private Hosting() {}

    /* ==================== profile ==================== */

    static final String TYPE_SFTP = "sftp";
    static final String TYPE_WEBDAV = "webdav";
    static final String TYPE_KUBO = "kubo";
    static final String TYPE_PINATA = "pinata";
    static final String TYPE_GITHUB = "github";
    static final String TYPE_ARWEAVE = "arweave";   // permanent, content-addressed (ArDrive Turbo)

    static class Profile {
        JSONObject j;

        Profile(JSONObject j) { this.j = j; }

        static Profile fresh(String type) {
            JSONObject o = new JSONObject();
            put(o, "v", 1);
            put(o, "id", "hp_" + System.currentTimeMillis());
            put(o, "name", "");
            put(o, "type", type);
            put(o, "def", false);
            put(o, "pathTemplate", "atelier/{collection}/{name}-{ts}{ext}");
            put(o, "dirTemplate", "atelier/{collection}-{ts}");
            put(o, type, new JSONObject());
            return new Profile(o);
        }

        String id() { return j.optString("id"); }
        String name() { return j.optString("name"); }
        String type() { return j.optString("type"); }
        boolean isDefault() { return j.optBoolean("def"); }
        JSONObject cfg() {
            JSONObject c = j.optJSONObject(type());
            return c == null ? new JSONObject() : c;
        }
        String cfgStr(String key) { return cfg().optString(key, ""); }
        String secret(String key) { return Crypt.decrypt(cfgStr(key)); }
        String pathTemplate() { return j.optString("pathTemplate", "atelier/{collection}/{name}-{ts}{ext}"); }
        String dirTemplate() { return j.optString("dirTemplate", "atelier/{collection}-{ts}"); }

        /** Hosts the user explicitly configured — allowed even on RFC1918. */
        java.util.HashSet<String> ownHosts() {
            java.util.HashSet<String> hs = new java.util.HashSet<>();
            for (String k : new String[]{ "host", "endpoint", "apiUrl", "gateway", "urlPrefix", "pagesPrefix" }) {
                String v = cfgStr(k);
                if (v.isEmpty()) continue;
                try {
                    hs.add(v.contains("://") ? new URL(v).getHost() : v);
                } catch (Exception ignored) { hs.add(v); }
            }
            return hs;
        }

        @Override public String toString() {   // redacted — never leaks secrets
            return "Profile{" + name() + " · " + type() + "}";
        }
    }

    /* ==================== pure helpers (parity with hosting.js) ============ */

    /** Slug: lowercase, [a-z0-9-], collapsed dashes, max 40, never empty. */
    static String slug(String s) {
        if (s == null) s = "";
        String out = java.text.Normalizer.normalize(s, java.text.Normalizer.Form.NFD)
                .replaceAll("\\p{M}", "")
                .toLowerCase(Locale.US)
                .replaceAll("[^a-z0-9]+", "-")
                .replaceAll("-+", "-")
                .replaceAll("^-|-$", "");
        if (out.length() > 40) out = out.substring(0, 40).replaceAll("-$", "");
        return out.isEmpty() ? "x" : out;
    }

    /** {collection} {name} {idx} {ext} {ts} {rand} — idx UNPADDED (the chain
     *  contract is base + i + ext with bare ints). Unknown tokens are left
     *  verbatim so mistakes are visible, not silent. */
    static String fillTemplate(String tpl, Map<String, String> tokens) {
        String out = tpl == null ? "" : tpl;
        for (Map.Entry<String, String> e : tokens.entrySet()) {
            out = out.replace("{" + e.getKey() + "}", e.getValue() == null ? "" : e.getValue());
        }
        return out;
    }

    static String ts36() {
        return Long.toString(System.currentTimeMillis() / 1000, 36);
    }

    static String rand4() {
        return Integer.toHexString((int) (Math.random() * 0xFFFF) | 0x10000).substring(1);
    }

    /** Public URL for an uploaded path (or CID for the IPFS backends).
     *  isDir=true returns a base ending in "/" ready for base+idx+ext. */
    static String publicUrl(Profile p, String relPathOrCid, boolean isDir) {
        String type = p.type();
        if (TYPE_ARWEAVE.equals(type)) {
            // relPathOrCid is a txid (single file) or manifest txid (directory)
            String gw = trimSlash(p.cfgStr("gateway"));
            if (gw.isEmpty()) gw = "https://arweave.net";
            return gw + "/" + relPathOrCid + (isDir ? "/" : "");
        }
        if (TYPE_KUBO.equals(type) || TYPE_PINATA.equals(type)) {
            String gw = trimSlash(p.cfgStr("gateway"));
            if (gw.isEmpty() && TYPE_PINATA.equals(type)) gw = "https://gateway.pinata.cloud";
            return gw + "/ipfs/" + relPathOrCid + (isDir ? "/" : "");
        }
        if (TYPE_GITHUB.equals(type)) {
            if ("pages".equals(p.cfgStr("serve"))) {
                return trimSlash(p.cfgStr("pagesPrefix")) + "/" + relPathOrCid + (isDir ? "/" : "");
            }
            return "https://raw.githubusercontent.com/" + p.cfgStr("owner") + "/" + p.cfgStr("repo")
                    + "/" + (p.cfgStr("branch").isEmpty() ? "main" : p.cfgStr("branch"))
                    + "/" + relPathOrCid + (isDir ? "/" : "");
        }
        // sftp/webdav: read-side prefix + path relative to the write root
        return trimSlash(p.cfgStr("urlPrefix")) + "/" + relPathOrCid + (isDir ? "/" : "");
    }

    static String trimSlash(String s) {
        if (s == null) return "";
        return s.endsWith("/") ? s.substring(0, s.length() - 1) : s;
    }

    /** kubo /api/v0/add NDJSON: one line per entry; the DIRECTORY entry's Name
     *  equals the shared prefix. Returns its Hash, or "" when absent. */
    static String parseKuboAddResponse(String ndjson, String dirName) {
        if (ndjson == null) return "";
        String last = "";
        for (String line : ndjson.split("\n")) {
            line = line.trim();
            if (line.isEmpty()) continue;
            try {
                JSONObject o = new JSONObject(line);
                String name = o.optString("Name");
                if (name.equals(dirName)) return o.optString("Hash");
                last = o.optString("Hash", last);
            } catch (Exception ignored) { }
        }
        return dirName == null || dirName.isEmpty() ? last : "";
    }

    /** GitHub contents-API body. Never carries a sha — that is the
     *  overwrite-refusal: GitHub 422s when the path already exists. */
    static String githubContentsBody(byte[] bytes, String message, String branch) {
        JSONObject o = new JSONObject();
        put(o, "message", message == null || message.isEmpty() ? "atelier upload" : message);
        put(o, "content", java.util.Base64.getEncoder().encodeToString(bytes));
        put(o, "branch", branch == null || branch.isEmpty() ? "main" : branch);
        return o.toString();
    }

    static String githubRawUrl(String owner, String repo, String branch, String path) {
        return "https://raw.githubusercontent.com/" + owner + "/" + repo + "/"
                + (branch == null || branch.isEmpty() ? "main" : branch) + "/" + path;
    }

    /* ==================== uploader surface ==================== */

    static class HostingException extends Exception {
        HostingException(String msg) { super(msg); }
    }

    interface ProgressCb { void onProgress(int done, int total, String msg); }

    static class Entry {
        final byte[] bytes; final String relPath; final String mime;
        Entry(byte[] bytes, String relPath, String mime) {
            this.bytes = bytes; this.relPath = relPath; this.mime = mime;
        }
    }

    /** Single-file transport. putFile returns the PUBLIC URL. */
    interface Uploader {
        String putFile(byte[] bytes, String relPath, String mime) throws HostingException;
        boolean exists(String relPath) throws HostingException;
    }

    /** Directory transport (IPFS backends): one call, one CID, base URL back. */
    interface DirUploader {
        String putDirectory(List<Entry> files, ProgressCb cb) throws HostingException;
    }

    static Uploader forProfile(Profile p) throws HostingException {
        switch (p.type()) {
            case TYPE_SFTP:   return new SftpUploader(p);
            case TYPE_WEBDAV: return new WebDavUploader(p);
            case TYPE_KUBO:   return new IpfsUploader(p);
            case TYPE_PINATA: return new PinataUploader(p);
            case TYPE_GITHUB: return new GithubUploader(p);
            case TYPE_ARWEAVE: return new ArweaveUploader(p);
            default: throw new HostingException("Unknown destination type: " + p.type());
        }
    }

    /** Sequential N-file upload for backends without directory semantics.
     *  Existence-checks EVERY target first (a renamed plate mid-batch would
     *  break the base+idx+ext contract), then uploads in order. Returns the
     *  common base URL (dir part) — caller derives base+idx+ext. */
    static String putSequence(Uploader u, Profile p, String dirRel, List<Entry> files, ProgressCb cb)
            throws HostingException {
        for (Entry e : files) {
            if (u.exists(e.relPath)) {
                throw new HostingException("Remote file already exists: " + e.relPath
                        + " — nothing was uploaded (files are never overwritten)");
            }
        }
        String lastUrl = "";
        int i = 0;
        for (Entry e : files) {
            if (cb != null) cb.onProgress(i, files.size(), e.relPath);
            lastUrl = u.putFile(e.bytes, e.relPath, e.mime);
            i++;
        }
        if (cb != null) cb.onProgress(files.size(), files.size(), "done");
        // base = last URL minus its filename
        int cut = lastUrl.lastIndexOf('/');
        return cut < 0 ? lastUrl : lastUrl.substring(0, cut + 1);
    }

    /* ==================== verification ==================== */

    /** The URL the mint will seal must serve — Range-GET so plates aren't
     *  pulled whole. SSRF policy: the standard blocklist applies EXCEPT for
     *  hosts the user explicitly configured in the profile (a LAN kubo
     *  gateway is legitimate for its owner). */
    /** Per-attempt progress hook for slow verifications. Return false to abort. */
    interface VerifyTick { boolean onTick(String url, int attempt, long elapsedMs); }

    static void verifyUrl(String url, Profile p) throws HostingException {
        verifyUrl(url, p, 0, null);
    }

    /** deadlineAt (epoch ms) caps the retry window for slow backends — pass one
     *  shared deadline when verifying a batch so N URLs can't each burn the full
     *  budget; 0 = per-call default. tick fires before every attempt. */
    static void verifyUrl(String url, Profile p, long deadlineAt, VerifyTick tick) throws HostingException {
        int attempts = 1;
        long sleepMs = 5000;
        boolean slow = p != null && TYPE_ARWEAVE.equals(p.type());
        if (p != null && TYPE_GITHUB.equals(p.type())) { attempts = 3; }   // raw CDN lag
        // Fresh Arweave txids 404 for ~5-10 min before the gateway indexes them
        // (measured live 2026-08-16) — retry until the deadline, return on first success.
        if (slow) {
            sleepMs = 15000;
            if (deadlineAt <= 0) deadlineAt = System.currentTimeMillis() + 10L * 60 * 1000;
            attempts = Integer.MAX_VALUE;   // deadline-bounded below
        }
        long start = System.currentTimeMillis();
        HostingException last = null;
        for (int a = 0; a < attempts; a++) {
            if (tick != null && !tick.onTick(url, a + 1, System.currentTimeMillis() - start)) {
                throw new HostingException("Verification stopped by user: " + url);
            }
            try {
                verifyOnce(url, p);
                return;
            } catch (HostingException e) {
                last = e;
                boolean more = slow ? System.currentTimeMillis() + sleepMs < deadlineAt : a < attempts - 1;
                if (!more) break;
                try { Thread.sleep(sleepMs); } catch (InterruptedException ignored) { }
            }
        }
        throw last;
    }

    private static void verifyOnce(String url, Profile p) throws HostingException {
        try {
            URL u = new URL(url);
            boolean own = p != null && p.ownHosts().contains(u.getHost());
            if (!own && ImageLoader.isBlockedHost(u.getHost())) {
                throw new HostingException("URL host is not publicly reachable: " + u.getHost());
            }
            HttpURLConnection con = (HttpURLConnection) u.openConnection();
            con.setConnectTimeout(8000);
            con.setReadTimeout(15000);
            con.setInstanceFollowRedirects(true);
            con.setRequestProperty("Range", "bytes=0-1");
            int code = con.getResponseCode();
            con.disconnect();
            if (code < 200 || code >= 400) {
                throw new HostingException("URL does not serve (HTTP " + code + "): " + url);
            }
        } catch (HostingException he) {
            throw he;
        } catch (Exception e) {
            throw new HostingException("URL not reachable: " + url + " — " + e.getClass().getSimpleName());
        }
    }

    /* ==================== HTTP plumbing shared by uploaders ============== */

    static HttpURLConnection open(String url, String method) throws Exception {
        HttpURLConnection con = (HttpURLConnection) new URL(url).openConnection();
        con.setConnectTimeout(10000);
        con.setReadTimeout(30000);
        con.setRequestMethod(method);
        return con;
    }

    static String readBody(HttpURLConnection con) {
        try (java.io.InputStream in = con.getResponseCode() < 400
                ? con.getInputStream() : con.getErrorStream()) {
            if (in == null) return "";
            java.io.ByteArrayOutputStream bos = new java.io.ByteArrayOutputStream();
            byte[] buf = new byte[8192];
            int n;
            while ((n = in.read(buf)) > 0) bos.write(buf, 0, n);
            return bos.toString("UTF-8");
        } catch (Exception e) {
            return "";
        }
    }

    static void put(JSONObject o, String k, Object v) {
        try { o.put(k, v); } catch (Exception ignored) { }
    }

    static void putArr(JSONArray a, Object v) {
        a.put(v);
    }
}
